group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,903 @@
+            if (!partitionMovementsByTopic.containsKey(topic))
+                partitionMovementsByTopic.put(topic, new HashMap<>());
+
+            Map<MemberPair, Set<TopicIdPartition>> partitionMovementsForThisTopic = partitionMovementsByTopic.get(topic);
+            if (!partitionMovementsForThisTopic.containsKey(pair))
+                partitionMovementsForThisTopic.put(pair, new HashSet<>());
+
+            partitionMovementsForThisTopic.get(pair).add(partition);
+        }
+
+        private void movePartition(TopicIdPartition partition, String oldOwner, String newOwner) {
+            MemberPair pair = new MemberPair(oldOwner, newOwner);
+
+            if (partitionMovementsByPartition.containsKey(partition)) {
+                // This partition was previously moved.
+                MemberPair existingPair = removeMovementRecordOfPartition(partition);
+                assert existingPair.dstMemberId.equals(oldOwner);
+                if (!existingPair.srcMemberId.equals(newOwner)) {
+                    // The partition is not moving back to its previous consumer.
+                    addPartitionMovementRecord(partition, new MemberPair(existingPair.srcMemberId, newOwner));
+                }
+            } else
+                addPartitionMovementRecord(partition, pair);
+        }
+
+        private TopicIdPartition getTheActualPartitionToBeMoved(
+            TopicIdPartition partition,
+            String oldOwner,
+            String newOwner
+        ) {
+            Uuid topic = partition.topicId();
+
+            if (!partitionMovementsByTopic.containsKey(topic))
+                return partition;
+
+            if (partitionMovementsByPartition.containsKey(partition)) {
+                String expectedOldOwner = partitionMovementsByPartition.get(partition).dstMemberId;
+                if (!oldOwner.equals(expectedOldOwner)) {
+                    throw new PartitionAssignorException("Old owner does not match expected value for partition: " + partition);
+                }
+                oldOwner = partitionMovementsByPartition.get(partition).srcMemberId;
+            }
+
+            Map<MemberPair, Set<TopicIdPartition>> partitionMovementsForThisTopic = partitionMovementsByTopic.get(topic);
+            MemberPair reversePair = new MemberPair(newOwner, oldOwner);
+            if (!partitionMovementsForThisTopic.containsKey(reversePair))
+                return partition;
+
+            return partitionMovementsForThisTopic.get(reversePair).iterator().next();
+        }
+    }
+
+    /**
+     * Manages assignments to members based on their current assignment size and maximum allowed assignment size.
+     */
+    private class AssignmentManager {
+        private final Map<String, MemberAssignmentData> membersWithAssignmentSizes = new HashMap<>();
+
+        /**
+         * Represents the assignment metadata for a member.
+         */
+        private class MemberAssignmentData {
+            final String memberId;
```

Comment:
oya thanks for the observation, I actually kept it because the sizes data corresponds to that of a particular member, if we want to print it or check equality it's better to have the memberId. What do you think?